src/qtcore/qml/lib/parser.js

```diff
@@ -293,20 +293,92 @@ function js_error(message, line, col, pos, comment) {
 function is_token(token, type, val) {
         return token.type == type && (val == null || token.value == val);
 };
+// CJK characters in the console display longer lengths, calculate the length of a string of actual display
+var normal_c = " ";
+var cjk_c = "　";
+function isCjk(char_code) {
+    return 0x4E00 <= char_code && char_code <= 0x9FFF ||
+        0x3400 <= char_code && char_code <= 0x4DFF ||
+        0x20000 <= char_code && char_code <= 0x2A6DF ||
+        0xF900 <= char_code && char_code <= 0xFAFF ||
+        0x2F800 <= char_code && char_code <= 0x2FA1F;
+};
+function fillWithWhitespace(str) {
+    var res = "";
+    for (var i = 0, len = str.length, cc; i < len; i += 1) {
+        cc = str.charCodeAt(i);
+        if (cc === 9) { // \t
+            res += '\t' // Ignore TAB, or else in a different console TAB length is not the same. ~~Unless there are modifications in the source code, the source TAB replaced by a fixed length spaces~~
+        } else if (isCjk(cc)) {
+            res += cjk_c
+        } else {
+            res += normal_c
+        }
+    }
+    return res;
+};
+// reference from [MDN | String/repeat](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/repeat).
+// Internal use only, ignoring the type of judgment, increase speed
+function repeatString(str, count) {
+    'use strict';
+    var rpt = '';
+    for (;;) {
+        if ((count & 1) == 1) {
+            rpt += str;
+        }
+        count >>>= 1;
+        if (count == 0) {
+            break;
+        }
+        str += str;
+    }
+    return rpt;
+};
```

Comment:
Btw, that code at MDN isn't actually good. It has an infinite loop and shift, is barely readable for the sake of optimizing performance, which I think it also fails to do.

In js, strings are immutable and pooled. Adding something to a string creates a new string in the pool, which is btw dependant on the previous string (because it consists of two parts — the old one plus the newly added one).

Constructing a string with concatenations of a single character is a good way to consume lots of memory in js =).

Though that's most likely not very important here (we don't expect lots of huge strings), I would just use `Array(count).fill(str).join('')` for the sake of readability.